starlette/staticfiles.py

```diff
@@ -158,20 +157,26 @@ async def get_response(self, path: str, scope: Scope) -> Response:
         raise HTTPException(status_code=404)
 
     def lookup_path(
-        self, path: str
-    ) -> typing.Tuple[str, typing.Optional[os.stat_result]]:
+        self, path: Path
+    ) -> typing.Tuple[Path, typing.Optional[os.stat_result]]:
         for directory in self.all_directories:
-            full_path = os.path.realpath(os.path.join(directory, path))
-            directory = os.path.realpath(directory)
-            if os.path.commonprefix([full_path, directory]) != directory:
+            original_path = Path(directory).joinpath(path)
+            full_path = original_path.resolve()
+            directory = Path(directory).resolve()
+            try:
+                full_path.relative_to(directory)
+                is_internal = True
+            except ValueError:
+                is_internal = False
+            if not is_internal and not original_path.is_symlink():
```

Comment:
Does is_symlink perform blocking IO?

Reply:
Can os.lstat be used so the stat result can be reused if the path isn't a symlink?

Reply:
> Does is_symlink perform blocking IO?

Not really sure about this.



> Can os.lstat be used so the stat result can be reused if the path isn't a symlink?

I think by that you mean on Line 171 we do `os.stat(full_path)` so you want to re-use that but here we check the symlink on the `original_path` before we resolve any symlinks, so they're different.
I might be missing something here....

Reply:
@graingert friendly reminder

Reply:
You can reuse the lstat if it turns out the original_path and full_path are the same

Reply:
Thanks for the review. Done. @graingert

Reply:
@graingert I don't fully understand what you mean by them being the same, if one is a link to another, how can they be the same?

Reply:
He's saying in case they are not.

Reply:
If they are not then we already have the correct lstat, so we're good.